File: Fingerprint-PY/crypto_handler.py
# crypto_handler.py
import hashlib
import oqs


class CryptoHandler:
    def __init__(self, sig_alg="Dilithium3"):
        self.sig_algorithm = sig_alg
        # 初始化签名处理器
        self.sig_handler = oqs.Signature(self.sig_algorithm)

    def sha256(self, data: bytes) -> bytes:
        """计算数据的SHA-256哈希"""
        return hashlib.sha256(data).digest()

    def sign(self, data_hash: bytes, private_key: bytes) -> bytes:
        """使用PQC私钥对哈希进行签名"""
        return self.sig_handler.sign(data_hash, private_key)

    def verify(self, data_hash: bytes, signature: bytes, public_key: bytes) -> bool:
        """使用PQC公钥验证签名"""
        try:
            return self.sig_handler.verify(data_hash, signature, public_key)
        except Exception:
            # oqs库在验证失败时可能会抛出异常
            return False

    def build_merkle_tree(self, leaf_hashes: list[bytes]) -> bytes:
        """根据叶子节点哈希列表构建Merkle树并返回根哈希"""
        if not leaf_hashes:
            return b''
        if len(leaf_hashes) == 1:
            return leaf_hashes[0]

        level = leaf_hashes
        while len(level) > 1:
            # 如果当前层节点数为奇数，复制最后一个节点以凑成偶数
            if len(level) % 2 != 0:
                level.append(level[-1])

            next_level = []
            for i in range(0, len(level), 2):
                # 将两个子节点的哈希拼接起来
                combined = level[i] + level[i + 1]
                # 计算父节点的哈希
                next_level.append(self.sha256(combined))

            level = next_level

        return level[0]

File: Fingerprint-C++/src/crypto_handler.cpp
// src/crypto_handler.cpp
#include "crypto_handler.hpp"
#include <openssl/evp.h>
#include <stdexcept>

CryptoHandler::CryptoHandler(const std::string& sig_alg) : signature_algorithm(sig_alg) {
    sig_handler = OQS_SIG_new(signature_algorithm.c_str());
    if (!sig_handler) { throw std::runtime_error("不支持的签名算法: " + sig_alg); }
}

CryptoHandler::~CryptoHandler() { if (sig_handler) { OQS_SIG_free(sig_handler); } }

std::vector<uint8_t> CryptoHandler::sha256(const std::string& data) const {
    std::vector<uint8_t> hash(EVP_MAX_MD_SIZE);
    unsigned int len;
    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    EVP_DigestInit_ex(ctx, EVP_sha256(), NULL);
    EVP_DigestUpdate(ctx, data.c_str(), data.length());
    EVP_DigestFinal_ex(ctx, hash.data(), &len);
    EVP_MD_CTX_free(ctx);
    hash.resize(len);
    return hash;
}

bool CryptoHandler::sign(const std::vector<uint8_t>& data, const std::vector<uint8_t>& sk, DeviceFingerprint::Signature* p) const {
    std::vector<uint8_t> sig(sig_handler->length_signature);
    size_t sig_len;
    if (OQS_SIG_sign(sig_handler, sig.data(), &sig_len, data.data(), data.size(), sk.data()) != OQS_SUCCESS) return false;
    sig.resize(sig_len);
    p->set_algorithm(signature_algorithm);
    p->set_signature_data(sig.data(), sig.size());
    return true;
}

bool CryptoHandler::verify(const std::vector<uint8_t>& data, const DeviceFingerprint::Signature& p, const std::vector<uint8_t>& pk) const {
    if (p.algorithm() != signature_algorithm) return false;
    std::string sig_data = p.signature_data();
    return OQS_SIG_verify(sig_handler, data.data(), data.size(), (uint8_t*)sig_data.c_str(), sig_data.size(), pk.data()) == OQS_SUCCESS;
}

std::vector<uint8_t> CryptoHandler::buildMerkleTree(const std::vector<std::vector<uint8_t>>& leaves) const {
    if (leaves.empty()) return {};
    if (leaves.size() == 1) return leaves[0];
    auto level = leaves;
    while (level.size() > 1) {
        if (level.size() % 2 != 0) level.push_back(level.back());
        std::vector<std::vector<uint8_t>> next_level;
        for (size_t i = 0; i < level.size(); i += 2) {
            std::string combined = std::string(level[i].begin(), level[i].end()) + std::string(level[i+1].begin(), level[i+1].end());
            next_level.push_back(sha256(combined));
        }
        level = next_level;
    }
    return level[0];
}
